### PR TITLE
C++: Fix FPs in cpp/badly-bounded-write caused by extraction errors

### PR DESCRIPTION
**Pull request by calumgrant**
#### All query authors

- [ ] A change note is added if necessary. See [the documentation](https://github.com/github/codeql/blob/main/docs/change-notes.md) in this repository.
- [ ] All new queries have appropriate `.qhelp`. See [the documentation](https://github.com/github/codeql/blob/main/docs/query-help-style-guide.md) in this repository.
- [ ] QL tests are added if necessary. See [Testing custom queries](https://docs.github.com/en/code-security/codeql-cli/using-the-advanced-functionality-of-the-codeql-cli/testing-custom-queries) in the GitHub documentation.
- [ ] New and changed queries have correct query metadata. See [the documentation](https://github.com/github/codeql/blob/main/docs/query-metadata-style-guide.md) in this repository.

#### Internal query authors only

- [ ] Autofixes generated based on these changes are valid, only needed if this PR makes significant changes to `.ql`, `.qll`, or `.qhelp` files. See [the documentation](https://github.com/github/codeql-team/blob/main/docs/best-practices/validating-autofix-for-query-changes.md) (internal access required).
- [ ] Changes are validated [at scale](https://github.com/github/codeql-dca/) (internal access required).
- [ ] Adding a new query? Consider also [adding the query to autofix](https://github.com/github/codeml-autofix/blob/main/docs/updating-query-support.md#adding-a-new-query-to-the-query-suite).

### CODE REVIEW

**Comment by calumgrant** [comment]
[DCA output](https://github.com/github/codeql-dca-main/blob/data/calumgrant/pr-18222-b7f47f__nightly-buildle__CustomSuite/reports/alert-comparison.md) looks good, showing the 8 FPs removed.

File: cpp/ql/src/Security/CWE/CWE-120/BadlyBoundedWrite.ql [hunk 1]
```diff
@@ -25,7 +25,8 @@ from BufferWrite bw, int destSize
 where
   bw.hasExplicitLimit() and // has an explicit size limit
   destSize = max(getBufferSize(bw.getDest(), _)) and
-  bw.getExplicitLimit() > destSize // but it's larger than the destination
+  bw.getExplicitLimit() > destSize and // but it's larger than the destination
+  not bw.getDest().getUnderlyingType().stripType() instanceof ErroneousType // destSize may be incorrect
```
[the remark this answers is not in the document]
Comment (jketema) [reply cpp/ql/src/Security/CWE/CWE-120/BadlyBoundedWrite.ql:29]: Yes, I believe `stripType` takes care of that already.

Comment (calumgrant) [reply cpp/ql/src/Security/CWE/CWE-120/BadlyBoundedWrite.ql:29]: I got confused, `stripType()` is only on `Type`.

Comment (jketema) [reply cpp/ql/src/Security/CWE/CWE-120/BadlyBoundedWrite.ql:29]: What's the type of `getDest`?

Comment (calumgrant) [reply cpp/ql/src/Security/CWE/CWE-120/BadlyBoundedWrite.ql:29]: It is `Expr`.

Comment (jketema) [reply cpp/ql/src/Security/CWE/CWE-120/BadlyBoundedWrite.ql:29]: You can probably just change this to `getType().stripType()` in that case.

**Comment by calumgrant** [comment]
I'll do another DCA once everything else looks good. (The custom suite ran just the query that changed.)

File: cpp/ql/src/change-notes/2024-12-05-badly-bounded-write.md [hunk 2]
```diff
@@ -0,0 +1,4 @@
+---
+category: minorAnalysis
+---
+* The "Badly bounded write" query (`cpp/badly-bounded-write`) query no longer produces results if there is an extraction error in the type of the output buffer.
```
Comment (jketema) [inline cpp/ql/src/change-notes/2024-12-05-badly-bounded-write.md:4]: Remove duplicate word
```suggestion
* The "Badly bounded write" query (`cpp/badly-bounded-write`) no longer produces results if there is an extraction error in the type of the output buffer.
```

**Comment by jketema** [comment]
This is now basically waiting for a more complete DCA run.

**Review by jketema — APPROVED** [review]

**Pull request merged**